### DISCUSSION TITLE
React 19 causing `useAtomValue` + `atomWithObservable` to re-suspense on new value

### DISCUSSION THREAD

**Discussion by tien** [pmndrs/jotai#2848] [Bug report]
This started happening with React 19 & I have to introduce a delay of 0 (queue macrotask) like so to resolve it:

```ts
useAtomValue(someAtom, { delay: 0 })
```

I will work on a shareable reproduction.

**Comment by dai-shi** [comment]
Thanks for reporting. Yeah, a repro with https://stackblitz.com would be helpful.

**Comment by dhruvkelawala** [comment]
Hello, do you have any updates for this issue? Seems like I am facing the same

**Comment by dhruvkelawala** [comment]
Not sure what `{ delay: 0 }` does as I couldn't find it in docs but it seemed to have worked. I am 100% sure now this is the issue I was facing. However, unable to create a reproduction with a simple example.

**Reply by dai-shi** [reply]
`{ delay: 0 }` is a hack of this kind of issues, but in a long term, we should look for a proper solution. It's fine to use it meanwhile.

> However, unable to create a reproduction with a simple example.

Hmm, at least a reproduction or ideally a test case is necessary for someone to fix...

**Reply by dhruvkelawala** [reply]
Yes, I'll keep trying. In our codebase,

```
import { liveQuery } from "dexie"
import { atomWithObservable } from "jotai/utils"

export const atomFromQuery = <T>(querier: () => T | Promise<T>) =>
  atomWithObservable(() => liveQuery(querier))
 ```
This causes the re-suspense issue without delay: 0.

**Reply by dhruvkelawala** [reply]
And it was not in react 18. Below is the example with react 19


https://github.com/user-attachments/assets/0cf7f29b-4e5c-4cac-a3f7-a0c52e02c869

**Comment by dhruvkelawala** [comment]
Hello, was there any update to this issue?

**Reply by dai-shi** [reply]
Is it possible to reproduce the bug without `dexie` dependency?
Your help would be necessary to create a failing test case.

**Reply by dhruvkelawala** [reply]
Hi, I spent yesterday trying to reproduce the issue. But on simple stackblitz app with/without dexie, I was unable to reproduce it. But it feels like if multiple atoms are nested and produces multiple promises, this can happen. However, I am really unsure how to reproduce it.

**Comment by tien** [comment]
@dai-shi I was finally able to reproduce it:

Issue happening on React 19: https://codesandbox.io/p/sandbox/loving-cache-djlkzp
Same issue doesn't happen for React 18: https://codesandbox.io/p/devbox/loving-cache-forked-vr3g3s

**Reply by dai-shi** [reply]
Thanks. I'll have a closer look later.

**Reply by tien** [reply]
If it help, I believe the issue happen because React 19 will always suspense for 300ms even if the promise is already fulfilled https://github.com/facebook/react/issues/31819

**Reply by dai-shi** [reply]
Yeah, that helps.

Here's the same repro with StackBlitz: https://stackblitz.com/edit/github-ytxnpgdy?file=src%2FApp.tsx

**Reply by dai-shi** [reply]
Minimized repro: https://stackblitz.com/edit/github-ytxnpgdy-yur3f7k5?file=src%2FApp.tsx

**Reply by dai-shi** [reply]
With https://github.com/pmndrs/jotai/pull/3013/commits/1685626fa4a68155c3fd76f7c3002c283d24f495, the repro keeps counting: https://stackblitz.com/edit/github-ytxnpgdy-xf6dcml2?file=package.json,src%2FApp.tsx

It still shows the loading with micro-suspense, but it's expected. The workaround is `{ delay: 0 }` in that case, or better not to use `async`. See also `jotai-derive`.

**Reply by tien** [reply]
Oh yeah, what I did in my application was to do both https://github.com/pmndrs/jotai/pull/3013 & use `jotai-derive`.

Idk what `attachPromiseMeta(createContinuablePromise(value))` does but it fixed all the issues for me. Using `delay: 0` also works, but it introduced another issue where `useTransition` no longer works with `useAtomValue` for some reason.

**Reply by dai-shi** [reply]
Hm, I can consider doing `attachPromiseMeta` always, but it requires a failing test without `atomWithObservable`. Can you try creating one?

**Reply by tien** [reply]
Yeah I'll try to create a test for it later once I'm back at my computer

**Reply by tien** [reply]
Also should I create an issue for this? Or is this expected?

> Using delay: 0 also works, but it introduced another issue where `useTransition` no longer works with `useAtomValue` for some reason.

**Reply by tien** [reply]
This is a separate discussion, but with React 19 changes to how suspend works, I feels like using promise is no longer optimal cuz of the fixed 300ms delay :( (but this might be something that React need to fix)

**Reply by dai-shi** [reply]
> Using delay: 0 also works, but it introduced another issue where `useTransition` no longer works with `useAtomValue` for some reason.

It works, but isn't in transition, right. It's totally expected, and it's why it's not the default behavior.

> I feels like using promise is no longer optimal

For our end, I'd recommend jotai-derive to avoid unnecessary suspense. Also, does using useTransition reduce triggering suspense?

**Reply by tien** [reply]
- Yeah I feel like for React 19 it's always best to use `jotai-derive` now.
- Not really, `useTransition` is just something that I noticed didn't work with `delay`.

**Reply by dai-shi** [reply]
> * Not really, `useTransition` is just something that I noticed didn't work with `delay`.

Can be something I'm not aware of. A minimal stackblitz reproduction would be helpful.

**Reply by tien** [reply]
So my problem was that `jotai-derive` doesn't work when mixing async & observable atoms:

```ts
import { derive } from "jotai-derive";

const someAtom = derive(
  [
    // `asyncAtom` will always be a promise, hence `someAtom` will always be a promise
    asyncAtom,
    observableAtom,
  ],
  (asyncValue, observableValue) => asyncValue + observableValue,
);
```

So I had to create my own async atom implementation

```ts
import { atom, type Getter } from "jotai";
import { startTransition } from "react";

type Data<T> = { value: T } | { error: unknown };

export function atomWithPromise<TValue>(
  getPromise: (
    get: Getter,
    options: { signal: AbortSignal },
  ) => TValue | Promise<TValue>,
) {
  const countAtom = atom(0);

  const promiseAtom = atom((get, { signal }) => {
    get(countAtom);

    const maybePromise = getPromise(get, { signal });

    const dataAtom = atom<Data<TValue | Promise<TValue>>>({
      value: maybePromise,
    });

    dataAtom.onMount = (update) => {
      if (maybePromise instanceof Promise) {
        void maybePromise
          .then((value) => startTransition(() => update({ value })))
          .catch((error) => startTransition(() => update({ error })));
      }
    };

    return dataAtom;
  });

  return atom(
    (get) => {
      const data = get(get(promiseAtom));

      if ("error" in data) {
        throw data.error;
      }

      return data.value;
    },
    (_, set) => {
      set(countAtom, (count) => count + 1);
    },
  );
}
```

**Reply by dai-shi** [reply]
Does `startTransition` do something special for your case?
Otherwise, it looks similar to unwarp or a combination of it. ref https://blog.axlight.com/posts/jotai-tips/#:~:text=Tip%2020%3A%20Async%20only%20initially (not exactly the same)

**Reply by tien** [reply]
If I don't add in those `startTransition`, `useTransition` elsewhere wouldn't work where `startTransition(() => refreshByIncrementingCount())` would causes 1 micro-suspense at the end for some reason.

I did try the `unwrap/loadable` approach before, and it also have the issue where `useTransition` no longer work.

**Reply by tien** [reply]
I feels like with React 19, promise are fundamentally no longer "cheap"/"free". So maybe "async only initially" should become the default new behaviour?

**Reply by dai-shi** [reply]
Maybe unwrap can take `startTransition` function. vanilla utils are not tied to React.
React recommendation might be like attachPromiseMeta on the library end, which I originally did but removed afterwards to make jotai/vanilla framework-agnostic.

**Reply by tien** [reply]
Yes that would help, but wouldn't exactly be "beginner friendly" imo, having to be aware of all the React 19 related hacks :p

**Reply by dai-shi** [reply]
Right..

**Reply by dai-shi** [reply]
I can reconsider #3013 as it may make sense for performance optimization.

**Reply by dai-shi** [reply]
#3026

**Reply by tien** [reply]
Been playing around with #3026 for a while now and it did eliminate all the unwanted suspense flickers for me, haven't noticed any adverse side-effect from turning that option on so far 👌

**Comment by stephanoskomnenos** [comment]
The most recent fix #3026 still needs both `delay: 0` and `unstable_promiseStatus: true` to work.
In the example using `jotai-cache`:
https://codesandbox.io/p/sandbox/jotai-obs-bug-19-forked-ld7rs3

**Reply by ebelanger** [reply]
Updating to the latest version of jotai (2.12.5) and adding `unstable_promiseStatus: true` seems to fix the problem.

**Reply by stephanoskomnenos** [reply]
> Updating to the latest version of jotai (2.12.5) and adding `unstable_promiseStatus: true` seems to fix the problem.

It seems I posted the wrong codesandbox link. Now updated.

**Reply by ebelanger** [reply]
You still need to use a startTransition in this scenario if you want to omit the delay, which I think is the standard way React concurrent rendering works.

**Reply by ebelanger** [reply]
From what I could trace, it seems the problem comes from `createContinuablePromise` which creates a pending promise although it receives an already fulfilled promise.  This promise is then passed to `React.use` which causes the micro-suspension.

This exemple would still need a startTransition though since switching to an uncached (and thus pending) value will cause suspension.

**Reply by stephanoskomnenos** [reply]
Thanks for your investigation. The problem is the cached results still cause suspension.

**Reply by dai-shi** [reply]
> From what I could trace, it seems the problem comes from `createContinuablePromise` which creates a pending promise although it receives an already fulfilled promise. This promise is then passed to `React.use` which causes the micro-suspension.

Does that mean, if we skip `createContinuablePromise`, the problem goes away?

**Reply by ebelanger** [reply]
> > From what I could trace, it seems the problem comes from `createContinuablePromise` which creates a pending promise although it receives an already fulfilled promise. This promise is then passed to `React.use` which causes the micro-suspension.
> 
> Does that mean, if we skip `createContinuablePromise`, the problem goes away?

After further investigation, it seems the problem comes from atomWithCache which returns a new promise.  So even though the result is cached, when atomWithCache computes, it generates a new pending promise which will cause suspension.

I'm no longer sure I understand how suspension works and I may be leading us into a false direction.  I apologize.

**Reply by dai-shi** [reply]
> atomWithCache which returns a new promise.

Oh, I see. atomWithCache stores the resolved value in the cache, so it creates a new promise with the resolved value. It might be technically possible to improve it by caching the promise itself.

**Comment by ebelanger** [comment]
I've been using `unstable_promiseStatus` and it seems to solve suspension triggering.  I've been wondering what are future plans for this option though.

Any information would be welcomed :)

**Reply by dai-shi** [reply]
> in a long term, we should look for a proper solution. It's fine to use it meanwhile.

is still valid. I think we keep `unstable_promiseStatus` until we see some changes for v3. https://github.com/pmndrs/jotai/discussions/2889#discussioncomment-12935507

**Comment by wojtekmaj** [comment]
I think this can be traced back to https://github.com/facebook/react/issues/31819. Whoever's affected by this, I'd suggest participating in this discussion.

**Comment by wojtekmaj** [comment]
*sigh* FINE, I'll fix it myself.

## Introducing [vite-plugin-react-fallback-throttle](https://www.npmjs.com/package/vite-plugin-react-fallback-throttle)

You can now use a Vite plugin to configure `FALLBACK_THROTTLE_MS` in Vite apps and Vitest unit tests.

### Usage

```ts
import { defineConfig } from 'vite';
import reactFallbackThrottlePlugin from 'vite-plugin-react-fallback-throttle';

export default defineConfig({
  plugins: [
    reactFallbackThrottlePlugin(), // Leave empty for 0, or provide your own value if you like
  ],
});
```

### Source

https://github.com/wojtekmaj/vite-plugin-react-fallback-throttle
